addons/l10n_es_edi_verifactu/models/verifactu_xml.py

```diff
@@ -0,0 +1,615 @@
+
+            recargo_percentage = recargo_equivalencia.get("tax_percentage")
+            recargo_amount = recargo_equivalencia.get("tax_amount")
+
+            detalle = {
+                'Impuesto': verifactu_tax_type,
+                'ClaveRegimen': clave_regimen,
+                'CalificacionOperacion': calificacion_operacion,
+                'OperacionExenta': exempt_reason,
+                'TipoImpositivo': self._format_number_Tipo2_2(tax_percentage),
+                'BaseImponibleOimporteNoSujeto': self._format_number_ImporteSgn12_2(base_amount_no_sujeto),
+                'BaseImponibleACoste': self._format_number_ImporteSgn12_2(base_amount_sujeto),
+                'CuotaRepercutida': self._format_number_ImporteSgn12_2(tax_amount),
+                'TipoRecargoEquivalencia': self._format_number_Tipo2_2(recargo_percentage),
+                'CuotaRecargoEquivalencia': self._format_number_ImporteSgn12_2(recargo_amount),
+            }
+
+            detalles.append(detalle)
+
+        total_amount = sign * (tax_details['base_amount'] + tax_details['tax_amount'])
+        tax_amount = sign * (tax_details['tax_amount'])
+
+        total_amount_formatted = self._format_number_ImporteSgn12_2(total_amount)
+        if total_amount_formatted != vals['identifier']['ImporteTotal']:
+            errors.append(_("The computed 'ImporteTotal' does not match the 'Total Signed' amount on the invoice."))
+
+        render_vals = {
+            'Macrodato': 'S' if abs(total_amount) >= 100000000 else None,
+            'Desglose': {
+                'DetalleDesglose': detalles,
+            },
+            'CuotaTotal': self._format_number_ImporteSgn12_2(tax_amount),
+            'ImporteTotal': total_amount_formatted,
+        }
+
+        return render_vals, errors
+
+    @api.model
+    def _get_db_identifier(self):
+        database_uuid = self.env['ir.config_parameter'].sudo().get_param('database.uuid')
+        return _sha256(database_uuid)
+
+    @api.model
+    def _render_vals_SistemaInformatico(self, vals):
+        errors = []
+
+        spanish_companies_on_db_count = self.env['res.company'].search_count([
+            ('account_fiscal_country_id.code', '=', 'ES'),
+        ], limit=2)
+
+        render_vals = {
+            'SistemaInformatico': {
+                'NombreRazon': 'ODOO ERP SP SL',
+                'NIF': 'B72659014',
+                'NombreSistemaInformatico': odoo.release.product_name,
+                'IdSistemaInformatico': '00',  # identifies Odoo the software as product of Odoo the company
+                'Version': odoo.release.version,
+                'NumeroInstalacion':  self._get_db_identifier(),
+                'TipoUsoPosibleSoloVerifactu': 'S',
+                'TipoUsoPosibleMultiOT': 'S',
+                'IndicadorMultiplesOT': 'S' if spanish_companies_on_db_count > 1 else 'N',
+            },
+        }
+
+        return render_vals, errors
+
+    @api.model
+    def _render_vals_dsig(self, vals):
```

Comment:
It seems like we have an entire logic to support JWS (json Web Signature) while we have the code for XML signature in e.g. Ticketbai.  And it is also the largest part of this file (and that could maybe go to a tools folder or something)

This file alone is 600 lines that I am still wondering if the XML would not be simpler.  (but for both, it should be possible somehow)